client/extensions/zoninator/components/settings/zone/posts-list.jsx

```diff
@@ -0,0 +1,90 @@
+/**
+ * External dependencies
+ */
+import React, { Component } from 'react';
+import { localize } from 'i18n-calypso';
+
+/**
+ * Internal dependencies
+ */
+import PostCard from './post-card';
+import RecentPostsDropdown from '../../recent-posts-dropdown';
+import SearchAutocomplete from './../../search-autocomplete';
+import FormFieldset from 'components/forms/form-fieldset';
+import SortableList from 'components/forms/sortable-list';
+
+class PostsList extends Component {
+
+	addPost = ( { push } ) => {
+		return ( post ) => push( post );
+	}
+
+	removePost = ( { remove } ) => {
+		return ( idx ) => remove( idx );
+	}
+
+	changePostOrder = ( { move } ) => {
+		return ( newOrder ) => {
+			if ( newOrder.length < 2 ) {
+				return;
+			}
+
+			// This loop attempts to find to which index in the array has been moved
+			// by making the following assumptions:
+			// Moved forward: newOrder[ from ] < from.
+			// Moved backward by less than one position: same as moving the next item forward.
+			// Moved backward by more than one position: newOrder[ from ] > from + 1.
+			let from = 0;
+
+			while ( ! ( newOrder[ from ] < from || newOrder[ from ] > from + 1 ) ) {
+				from += 1;
+			}
```

Comment:
Hmm, this algorithm looks a bit more complex than necessary, given the available information and `FieldArray`'s [methods](http://redux-form.com/7.0.3/docs/api/FieldArray.md/). I think I'm especially wary about mutating `from` inside a loop that only ends based on a criterion for that `from`; I'd prefer something that doesn't involve mutating, or an possibly unterminated `while` loop :smile: Maybe we can find a simpler way by ways of a couple of assumptions :thinking: 

* `changePostOrder` will be called upon each change, so it's save to assume that only one item has been moved.
* Trying to come up with a representative example: Let's say `newOrder` is `0, 1, 6, 3, 4, 5, 2, 7`.
* So I think we'd basically just need to find the index that doesn't equal its value (which is the new index).

```js
const from = findindex( newOrder, ( newIndex, index ) => ( newIndex !== index ) );
```

Would that work?

Reply:
Not exactly. For example, take an array with 4 items: `0, 1, 2, 3`.
Now if we move `1` to the back, `newOrder` becomes: `0, 2, 3, 1`.
And if we were to move `3` to be the second item: `0, 3, 1, 2`.

Also, note that moving `1` to `3` and `3` to `1` are not equivalent.

Reply:
Hmm, I'll stop spitballing now -- but it'd be really great to find something non-mutating, non-potentially-unterminated :slightly_smiling_face:

Reply:
So just checking for inequality isn't enough, we also need to know if it's because of a gap (`from`)  or because of an intruder ( `to`)...

Reply:
Which I guess your original algorithm does just fine. Can we just plug your `while` criterion into   `findIndex`? Bonus points for using the current value, i.e. `newIndex`, instead of `newOrder[ index ]` 😄

Reply:
That won't work as I need the value as well as its index, `findIndex` only gives me access to the value.